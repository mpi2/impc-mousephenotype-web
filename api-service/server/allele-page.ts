import { fetchAPIFromServer, fetchURL } from "@/api-service";
import { AlleleSummary } from "@/models";
const KUBERNETES_NAMESPACE = process.env.KUBERNETES_NAMESPACE ?? "default";
const NODE_ENV = process.env.NODE_ENV;

export async function fetchAlleleSummary(
  mgiGeneAccessionId: string,
  alleleName: string,
): Promise<AlleleSummary> {
  const endpointURL = `http://impc-allele-service.${KUBERNETES_NAMESPACE}:8080/v1/alleles?mgiGeneAccessionId=${mgiGeneAccessionId}&alleleName=${alleleName}`;
  return await (NODE_ENV === "development"
    ? fetchAPIFromServer<AlleleSummary>(
        `/api/v1/alleles/${mgiGeneAccessionId}/${alleleName}`,
      )
    : fetchURL<AlleleSummary>(endpointURL));
}
