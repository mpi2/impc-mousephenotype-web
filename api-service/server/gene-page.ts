import { GeneOrder, GenePhenotypeHits, GeneSummary } from "@/models/gene";
import { fetchAPIFromServer, fetchURL } from "@/api-service";
import { processGeneOrderResponse } from "@/hooks/gene-order.query";
import { processGenePhenotypeHitsResponse } from "@/hooks/significant-phenotypes.query";

const KUBERNETES_NAMESPACE = process.env.KUBERNETES_NAMESPACE ?? "default";
const NODE_ENV = process.env.NODE_ENV;

export async function fetchGeneSummary(
  mgiGeneAccessionId: string,
): Promise<GeneSummary> {
  const endpointURL = `http://impc-summary-service.${KUBERNETES_NAMESPACE}:8080/v1/gene_summary?mgiGeneAccessionId=${mgiGeneAccessionId}`;
  return await (NODE_ENV === "development"
    ? fetchAPIFromServer<GeneSummary>(
        `/api/v1/genes/${mgiGeneAccessionId}/summary`,
      )
    : fetchURL<GeneSummary>(endpointURL));
}

export async function fetchGenePhenotypeHits(
  mgiGeneAccessionId: string,
): Promise<Array<GenePhenotypeHits>> {
  const endpointURL = `http://impc-phenotype-hits-service.${KUBERNETES_NAMESPACE}:8080/v1/phenotypehits?mgiGeneAccessionId=${mgiGeneAccessionId}`;
  const data = await (NODE_ENV === "development"
    ? fetchAPIFromServer<Array<GenePhenotypeHits>>(
        `/api/v1/genes/${mgiGeneAccessionId}/phenotype-hits`,
      )
    : fetchURL<Array<GenePhenotypeHits>>(endpointURL));
  return processGenePhenotypeHitsResponse(data);
}

export async function fetchGeneOrderData(
  mgiGeneAccessionId: string,
): Promise<Array<GeneOrder>> {
  const endpointURL = `http://impc-order-service.${KUBERNETES_NAMESPACE}:8080/v1/orders?mgiGeneAccessionId=${mgiGeneAccessionId}`;
  const data = await (NODE_ENV === "development"
    ? fetchAPIFromServer<Array<GeneOrder>>(
        `/api/v1/genes/${mgiGeneAccessionId}/order`,
      )
    : fetchURL<Array<GeneOrder>>(endpointURL));
  return processGeneOrderResponse(data);
}
