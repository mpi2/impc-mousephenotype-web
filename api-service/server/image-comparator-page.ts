import { GeneImageCollection } from "@/models/gene";
import { fetchAPIFromServer, fetchURL } from "@/api-service";

const KUBERNETES_NAMESPACE = process.env.KUBERNETES_NAMESPACE ?? "default";
const NODE_ENV = process.env.NODE_ENV;

export async function fetchMutantImages(
  mgiGeneAccessionId: string,
  parameterStableId: string,
): Promise<Array<GeneImageCollection>> {
  const endpointURL = `http://impc-imaging-service.${KUBERNETES_NAMESPACE}:8080/v1/images/findByMgiAndStableId?mgiGeneAccessionId=${mgiGeneAccessionId}&parameterStableId=${parameterStableId}`;
  try {
    return await (NODE_ENV === "development"
      ? fetchAPIFromServer<Array<GeneImageCollection>>(
          `/api/v1/images/find_by_mgi_and_stable_id?mgiGeneAccessionId=${mgiGeneAccessionId}&parameterStableId=${parameterStableId}`,
        )
      : fetchURL<Array<GeneImageCollection>>(endpointURL));
  } catch (error) {
    return [];
  }
}

export async function fetchControlImages(
  parameterStableId: string,
): Promise<Array<GeneImageCollection>> {
  const endpointURL = `http://impc-imaging-service.${KUBERNETES_NAMESPACE}:8080/v1/images/findByStableIdAndSampleId?biologicalSampleGroup=control&parameterStableId=${parameterStableId}`;
  try {
    return await (NODE_ENV === "development"
      ? fetchAPIFromServer<Array<GeneImageCollection>>(
          `/api/v1/images/find_by_stable_id_and_sample_id?biologicalSampleGroup=control&parameterStableId=${parameterStableId}`,
        )
      : fetchURL<Array<GeneImageCollection>>(endpointURL));
  } catch (error) {
    return [];
  }
}
