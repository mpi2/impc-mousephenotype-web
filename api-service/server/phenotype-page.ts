import { fetchAPIFromServer, fetchURL } from "@/api-service";
import { PhenotypeSummary } from "@/models/phenotype";

const KUBERNETES_NAMESPACE = process.env.KUBERNETES_NAMESPACE ?? "default";
const NODE_ENV = process.env.NODE_ENV;

export async function fetchPhenotypeSummary(
  phenotypeId: string,
): Promise<PhenotypeSummary> {
  const endpointURL = `http://impc-summary-service.${KUBERNETES_NAMESPACE}:8080/v1/phenotype_summary?phenotypeId=${phenotypeId}`;
  return await (NODE_ENV === "development"
    ? fetchAPIFromServer<PhenotypeSummary>(
        `/api/v1/phenotypes/${phenotypeId}/summary`,
      )
    : fetchURL<PhenotypeSummary>(endpointURL));
}
