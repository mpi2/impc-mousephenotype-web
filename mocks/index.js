export async function initMocks() {
  if (typeof window === "undefined") {
    const { server } = await import("./server");
    server.listen();
  } else {
    const { worker } = await import("./browser");
    return worker.start({
      serviceWorker: {
        url: "/data/mockServiceWorker.js",
        options: {
          scope: "/",
        }
      }
    });
  }
}

