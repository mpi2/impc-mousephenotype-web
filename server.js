const { createServer } = require('http');
const { parse } = require('url');
const next = require('next');

const dev = process.env.NODE_ENV !== 'production';
const hostname = 'localhost';
const port = 3000;
const app = next({ dev, hostname, port });
const handle = app.getRequestHandler();

app.prepare()
.then(() => {
    createServer(async (req, res) => {
        try {
            const parsedUrl = parse(req.url, true);
            const { pathname, query} = parsedUrl;
            const host = req.headers.host;
            console.log({ host, pathname });
            res.setHeader('X-IMPC-TEST', "true");
            if (host.includes('nginx.mousephenotype-dev.org')) {
                const modifiedPathName = pathname.replace("/data/", "");
                console.log({ modifiedPathName });
                await app.render(req, res, modifiedPathName, query);
                res.setHeader('X-PATH', JSON.stringify({ host, pathname, modifiedPathName }));
            } else {
                await handle(req, res, parsedUrl);
            }

        } catch (err) {
            console.error('Error occurred handling', req.url, err)
            res.statusCode = 500
            res.end('internal server error')
        }
    })
        .once('error', (err) => {
            console.error(err);
            process.exit(1);
        })
        .listen(port, () => {
            console.log(`Ready on http://${hostname}:${port}`);
        })
});