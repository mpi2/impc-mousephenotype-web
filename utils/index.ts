import _ from "lodash";

export const formatBodySystems = (systems: string[] | string = []) => {
  return _.capitalize(
    (typeof systems === "string" ? systems : systems.join(", "))
      .replace(/ phenotype/g, "")
      .replace(/\//g, " / ")
  );
};

export const formatAlleleSymbol = (allele: string) => {
  return allele.slice(0, allele.length - 1).split("<");
};
