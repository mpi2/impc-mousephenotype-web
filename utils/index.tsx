import React from "react";
import _ from "lodash";

export const formatBodySystems = (systems: string[] | string = []) => {
  return _.capitalize(
    (typeof systems === "string" ? systems : systems.join(", "))
      .replace(/ phenotype/g, "")
      .replace(/\//g, " / ")
  );
};

export const formatAlleleSymbol = (allele: string) => {
  return allele.slice(0, allele.length - 1).split("<");
};

export const formatPValue = (pValue: number) => {
  const pValueArray = Number.parseFloat(String(pValue))
    .toExponential(2)
    .split("e");
  return (
    <>
      {pValueArray[0]}x10<sup>{pValueArray[1].replace("+", "")}</sup>
    </>
  );
};

export const formatESCellName = (src: string) => {
  const [name, superscript, end] = src.split(/<|>/);
  return (
    <>
      {name}
      <sup>{superscript}</sup>
      {end}
    </>
  );
};

const fetchCache = {};
export { fetchCache };

export const toSentenceCase = (camelCase) => {
  if (camelCase) {
    const result = camelCase.replace(/([A-Z])/g, " $1");
    return result[0].toUpperCase() + result.substring(1).toLowerCase();
  }
  return "";
};
