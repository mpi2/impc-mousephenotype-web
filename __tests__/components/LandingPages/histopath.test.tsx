import { within, screen, waitFor } from "@testing-library/react";
import HistopathLandingPage from "@/app/histopath/histopath-page";
import { renderWithClient } from "../../utils";
import userEvent from "@testing-library/user-event";
import { testServer } from "../../../mocks/server";
import { rest } from "msw";
import histopathologyData from "../../../mocks/data/tests/landing-pages/histopathology.json";

jest.mock("next/navigation", () => ({
  useRouter: jest.fn(),
  useSearchParams: jest.fn().mockImplementation(() => new URLSearchParams()),
  usePathname: jest.fn(),
}));

describe("Histopathology landing page", () => {
  testServer.use(
    rest.get(
      `https://impc-datasets.s3.eu-west-2.amazonaws.com/landing-page-data/dr22.1/histopathology_landing.json`,
      (req, res, ctx) => {
        return res(ctx.json(histopathologyData));
      },
    ),
  );
  it("renders correctly", async () => {
    const user = userEvent.setup();
    const { container } = renderWithClient(<HistopathLandingPage />);
    await waitFor(() => expect(screen.getByRole("table")).toBeInTheDocument());
    expect(
      await screen.findByText("55", { selector: "span" }),
    ).toBeInTheDocument();
    expect(container).toMatchSnapshot();
    await user.type(screen.getByTitle("gene search box"), "Rpl12");
    await waitFor(async () => {
      const rows = await screen.findAllByRole("row");
      return expect(rows.length).toEqual(3);
    });
    await waitFor(async () => {
      const rows = await screen.findAllByRole("columnheader");
      return expect(rows.length).toEqual(106);
    });
    expect(await screen.findByTitle("fixed tissue link")).toHaveAttribute(
      "href",
      "/data/alleles/MGI:98002/em1(IMPC)Mbp#mice",
    );
    await user.clear(screen.getByTitle("gene search box"));
    await user.click(screen.getByTestId("bone-header"));
    let rows = await screen.findAllByTestId("result-rows");
    expect(within(rows[0]).getByTestId("gene-symbol")).toHaveTextContent(
      "Col9a3",
    );
    await user.click(screen.getByTestId("knee-joint-header"));
    rows = await screen.findAllByTestId("result-rows");
    expect(within(rows[0]).getByTestId("gene-symbol")).toHaveTextContent(
      "Chsy3",
    );
    await user.click(screen.getByTestId("gene-symbol-sort"));
    rows = await screen.findAllByTestId("result-rows");
    expect(within(rows[0]).getByTestId("gene-symbol")).toHaveTextContent(
      "Ache",
    );
  }, 10000);
});
