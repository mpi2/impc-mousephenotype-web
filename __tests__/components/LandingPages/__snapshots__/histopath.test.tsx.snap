// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Histopathology landing page renders correctly 1`] = `
<div>
  <div
    class="banner"
  >
    <div
      class="pb-4 pt-5 container container"
    >
      <div
        class="col-12 col-md-8 ps-4 pe-4"
      >
        <div
          class="tabs"
        >
          <a
            class="tab__active"
            href="/search"
          >
            Genes
          </a>
          <a
            class="tab"
            href="/search?type=pheno"
          >
            Phenotypes
          </a>
          <a
            class="tab"
            href="https://dev.mousephenotype.org/?s="
          >
            Help, news, blog
          </a>
        </div>
        <div
          class="inputCont"
        >
          <input
            class="input"
            placeholder="Search for a gene..."
            title="main search box"
            type="text"
            value=""
          />
          <button
            aria-describedby="svg-inline--fa-title-search-icon"
            class="searchBtn"
          >
            <svg
              aria-labelledby="svg-inline--fa-title-search-icon"
              class="svg-inline--fa fa-magnifying-glass "
              data-icon="magnifying-glass"
              data-prefix="fas"
              role="img"
              viewBox="0 0 512 512"
              xmlns="http://www.w3.org/2000/svg"
            >
              <title
                id="svg-inline--fa-title-search-icon"
              >
                Search button
              </title>
              <path
                d="M416 208c0 45.9-14.9 88.3-40 122.7L502.6 457.4c12.5 12.5 12.5 32.8 0 45.3s-32.8 12.5-45.3 0L330.7 376c-34.4 25.2-76.8 40-122.7 40C93.1 416 0 322.9 0 208S93.1 0 208 0S416 93.1 416 208zM208 352a144 144 0 1 0 0-288 144 144 0 1 0 0 288z"
                fill="currentColor"
              />
            </svg>
          </button>
        </div>
      </div>
    </div>
  </div>
  <div
    class="page container"
    style="line-height: 2;"
  >
    <div
      class="card"
    >
      <div
        class="subheading"
      >
        <nav
          aria-label="breadcrumb"
        >
          <ol
            class="breadcrumb"
          >
            <li
              aria-current="page"
              class="breadcrumb-item active"
            >
              Home
            </li>
            <li
              aria-current="page"
              class="breadcrumb-item active"
            >
              IMPC data collections
            </li>
            <li
              aria-current="page"
              class="breadcrumb-item active"
            >
              Histopathology Data
            </li>
          </ol>
        </nav>
      </div>
      <h1
        class="mb-4 mt-2"
      >
        <strong>
          Histopathology Data
        </strong>
      </h1>
      <div
        class="container"
      >
        <div>
          <p
            class="my-0"
          >
            <b>
              Significance Score:
            </b>
          </p>
          <div>
            <div
              class="labelContainer"
              title="No Data"
            >
              <div
                class="heatmapLabel"
                style="background-color: rgb(255, 255, 255);"
              />
              <div>
                  No Data
              </div>
            </div>
            <div
              class="labelContainer"
              title="Not Applicable"
            >
              <div
                class="heatmapLabel"
                style="background-color: rgb(128, 128, 128);"
              />
              <div>
                  Not Applicable
              </div>
            </div>
            <div
              class="labelContainer"
              title="Not Significant"
            >
              <div
                class="heatmapLabel"
                style="background-color: rgb(23, 162, 184);"
              />
                
              <div>
                <b>
                  Not Significant
                </b>
                 (histopathology finding that is interpreted by the histopathologist to be within normal limits of background strain-related findings or an incidental finding not related to genotype)
              </div>
            </div>
            <div
              class="labelContainer"
              title="Significant"
            >
              <div
                class="heatmapLabel"
                style="background-color: rgb(206, 98, 17);"
              />
                
              <div>
                <b>
                  Significant
                </b>
                 (histopathology finding that is interpreted by the histopathologist to not be a background strain-related finding or an incidental finding)
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        class="topControls"
      >
        <div>
          Gene search:
          <input
            class="form-control"
            title="gene search box"
            type="text"
            value=""
          />
        </div>
        <div>
          Show
          <select
            class="form-select"
            name="pageSize"
          >
            <option
              value="10"
            >
              10
            </option>
            <option
              value="25"
            >
              25
            </option>
            <option
              value="50"
            >
              50
            </option>
            <option
              value="100"
            >
              100
            </option>
          </select>
          entries
        </div>
      </div>
      <div
        class="mt-4"
        style="display: flex; justify-content: center;"
      >
        <div
          data-test-id="CircularProgressbarWithChildren"
        >
          <div
            style="position: relative; width: 100%; height: 100%;"
          >
            <svg
              class="CircularProgressbar rotating"
              data-test-id="CircularProgressbar"
              viewBox="0 0 100 100"
            >
              <path
                class="CircularProgressbar-trail"
                d="
      M 50,50
      m 0,-44
      a 44,44 0 1 1 0,88
      a 44,44 0 1 1 0,-88
    "
                fill-opacity="0"
                stroke-width="12"
                style="stroke: #fff; stroke-dasharray: 276.46015351590177px 276.46015351590177px; stroke-dashoffset: 0px;"
              />
              <path
                class="CircularProgressbar-path"
                d="
      M 50,50
      m 0,-44
      a 44,44 0 1 1 0,88
      a 44,44 0 1 1 0,-88
    "
                fill-opacity="0"
                stroke-width="12"
                style="stroke: #ccc; transition: stroke-dashoffset 2s ease 0s; stroke-dasharray: 276.46015351590177px 276.46015351590177px; stroke-dashoffset: 55.29203070318034px;"
              />
            </svg>
          </div>
        </div>
      </div>
      <div
        class="mt-1"
      >
        <span>
          * This column is from data made available to us, send us a message to inquiry about genes with no fixed tissue from 
          <a
            class="link primary"
            href="http://www.mousephenotype.org/contact-us/"
          >
            our Contact page
          </a>
        </span>
      </div>
      <div
        class="mt-1"
      />
    </div>
  </div>
</div>
`;
